source/fundamentals/promises.txt

```diff
@@ -0,0 +1,194 @@
+======================
+Promises and Callbacks
+======================
+
+.. default-domain:: mongodb
+
+.. contents:: On this page
+   :local:
+   :backlinks: none
+   :depth: 1
+   :class: singlecols
+
+Overview
+--------
+
+The NodeJS driver uses the asynchronous Javascript API to communicate with
+your MongoDB cluster.
+
+Asynchronous Javascript allows you to execute operations without waiting for
+the processing thread to become free. This helps prevent your application
+from becoming unresponsive to web UI rendering or network requests when
+executing long-running operations. For more information about asynchronous
+Javascript, see the MDN web documentation on
+`Asynchronous Javascript <https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Asynchronous>`_.
+
+This section describes two features of asynchronous Javascript --
+`Promises`_ and `Callbacks`_ -- that you can use with the NodeJS driver to
+access the results of your method calls to your MongoDB cluster.
+
+Promises
+--------
+
+A Promise is an object returned by the asynchronous method call that allows
+you to access information on the eventual success or failure of the operation.
+They are often used to execute operations that take some time and have
+when you want to notify once it's completed. For more information on Promises
+and related terminology, see the MDN documentation on
+:mdn:`Promises <Web/JavaScript/Reference/Global_Objects/Promise>`.
+
+If you call a driver method such as ``findOneAndModify()``,
+``countDocuments()``, or ``update()``, the NodeJS driver returns a Promise.
+
+You can use the ``await`` operator inside of an ``async`` method on the
+Promise to cause the enclosing method to pause until the Promise reaches
+either the **Fulfilled** or **Rejected** state. If it reaches the **Fulfilled**
+state, the reference contains the operation result as an object or value.
+
+.. code-block:: js
+
+   async function run() {
+     ...
+     const result = await collection.insertOne({ name: "Mount Everest", meters: 8848 });
+
+     // print the result of the insertOne() call
+     console.log(result);
+   }
+
+If you want to execute logic once the Promise reaches the **Fulfilled**
+or **Rejected** state, you can append the ``then()`` method to the Promise.
+The first parameter is the method that gets called when the Promise reaches
+the **Fulfilled** state and the optional second parameter is the method that
+gets called when it reaches the **Rejected** state. These methods are called
+**callbacks** because they execute only after the Promise completes.
+
+.. code-block:: js
+
+   ...
+   collection.updateOne({ name: "Mount McKinley" }, { meters: 6190 })
+     .then(
+       res => console.log(`Updated ${res.result.n} documents`),
+       err => console.error(`Something went wrong: ${err}`)
```

Comment:
? I believe if the operation errors it will end up being an unhandled promise rejection in this example because there is no `.catch()` chain. Am I mistaken?

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise#Chained_Promises

Reply:
This example is demonstrating using both at the same time. The examples you linked omit the catch in each of the `then()` methods, and then catch them at the end which is a different use case. I can show that too, if it's necessary.